### ISSUE TITLE
Allow to configure title bar information

### ISSUE THREAD

**Issue by alexdresko** [Microsoft/vscode#1723]
It would be great if we could configure the information that gets displayed in the title bar. I use ManicTime a lot, and it's hard to auto tag my activities when the title bar basically only displays the file name and root folder name. Having the full path to the file and the git branch name would be super helpful. 

For some other ideas, this VS plugin is super rad and affords complete control over what gets displayed in the title bar. https://visualstudiogallery.msdn.microsoft.com/2e8ebfe4-023f-4c4d-9b7a-d05bbc5cb239

**Comment by alexdresko** [comment]
@bpasero By assigning this to yourself, does that mean it might happen? If so, please let me know if you would like some help, even if it's just for feedback.

**Comment by bpasero** [comment]
This issue is currently assigned to the backlog. Every sprint we pick items from the backlog to plan for the current iteration. Please see https://github.com/Microsoft/vscode/wiki/Issue-Tracking#planning

**Comment by jamietre** [comment]
+1. The more I use VS Code on bigger projects, it's becoming frustrating to be unable to distinguish between open windows.  I wish I could configure the title to just show [Root Folder] - [File], since it's hard to distinguish between open projects when the file name is shown first. But some simple templating system would probably solve everyone's problem.

**Comment by bpasero** [comment]
https://github.com/Microsoft/vscode/issues/12625 for the issue to show the full path in the title bar. This might resolve many (but not all) concerns of this issue.

**Comment by jamietre** [comment]
@bpasero i do appreciate the effort but actually this still does not address the issue (at least as far as Windows) is concerned. It may actually make it worse (though since it's an option that's not an issue).

The problem is that in Windows, the taskbar shows a preview of all windows  for a given application, and there is only limited space to show the title. Showing the full path would give me exactly the opposite problem I have now: I'd just see the part of the path, starting with the root, that fits in the display. So with this enabled, I'd most likely have no information at all about which window was which, since they likely all share a common root path.

What is *really* needed is the ability to show `project-root/../filename`. The parent folder is nearly always the identifying feature for a project. I don't even care if the filename gets cut off- because I really just want to be able to determine which project is open from the title bar.

![image](https://cloud.githubusercontent.com/assets/573056/19929754/4e083920-a0db-11e6-810a-24f161bfcd71.png)

**Comment by bpasero** [comment]
@jamietre we can still tweak this experience and I welcome the problem description you gave, though I am not sure how to best solve it.

Here is how your scenario looks like without the new option enabled:

![image](https://cloud.githubusercontent.com/assets/900690/19940313/87818d9a-a12c-11e6-9611-5ad2aea0cf97.png)

And here with the option turned on:

![image](https://cloud.githubusercontent.com/assets/900690/19940343/9c3f3192-a12c-11e6-8b6c-60b79ceed3ba.png)

This makes it hard to see the currently opened file but you can easily see the full path when you compare the 2 window titles in full (not from this experience).

I think this feature request is still valid for that particular scenario because you could then really compose the title you want, I am not sure we should change the experience by default with the new option.

I am open for suggestions though if the default experience when this new option is turned on should be changed.

**Comment by jamietre** [comment]
Thanks for all the feedback, it's very appreciated. 

In my case almost all my code is under a single heirarchy like `d:/users/jtreworgy/code/...` -- so if I were lucky I might get a few characters of the root folder. But I also work in subfolders with groups of projects, too, so it would be quite realistic for there to be nothing unique at all. I could try to keep everything under shallow heirarchies but I guess I'd rather not have to design my file system architecture around accomodating VS Code ;)

The perfect world for me would be the ability to pass a template like `"{rootFolder} - {fileName}`" instead of boolean options for predefined templates. Just having `rootFolder`, `fileName` and `fullPath` as available parameters would accomodate everything that's been discussed here as well as the current state. 

Even more perfect world, the ability to invoke a  javascript function that I wrote to return my title given the path as input :) but then I've written a plugin haven't I..

**Comment by bpasero** [comment]
@jamietre but how is root folder solving it when you have 2 workspaces with the same name in different hierarchies?

"{rootFolder} - {fileName}" isn't that what we are showing today without the full path option enabled?

**Comment by jamietre** [comment]
I wasn't considering that possibility - I would never have two workspaces with the same name open (or at least, it's not particularly likely, maybe if I was trying to deep dive some comparison between two branches?). But that's not my problem generally... if that is an issue for some people, perhaps a possible solution is to also display a numerical index  identifying which was opened first? Can VS code instances know anything about each other?

But anyway -- right now the problem is the root folder appears *last*, e.g. `{fileName} - {rootFolder}` so it is cut off for long file names (which are pretty common in our projects). And it's hard to parse them even if they are visble when they aren't left-aligned.

**Comment by bpasero** [comment]
Oh yeah, that is true. I was not thinking about the fact that the file name comes first. We cannot change that without making it hard for users that want to quickly glance at the file name in that scenario so I guess no matter what it boils down to making it configurable if any.

I still think the setting to show a full path is a good one, so I will keep it as is for now and we leave this one here open.

**Comment by Cedric-Venet** [comment]
I would really love to hve an option/config to at least show {rootFolder} first. event if its just that (don't really see the point of having the filename in the title bar. For me the next most useful information would be the git branch name: "{rootFolder} - {branchName}"

**Comment by jamietre** [comment]
@Cedric-Venet agreed. If VS Code allowed you to drag a file out to start a new window, I could see how that might be useful to some. But you can't -- so it just doesn't seem likely that's what many people are looking for a file name when trying to discriminate between open instances. 

I looked at the code that creates the window template it `vs/workbench/services/history/browser/history.ts`... it seems straightforward and I am sure I could handle the change to support user-provided templates if this change would be accepted. Would you be interested in a PR? I have no idea what the overhead is in getting set up to create vs code builds - if you can convince me it's not too bad I'm willing to take a shot :)

**Comment by alexdresko** [comment]
@jamietre Dunno if this exactly deals with one of your use cases, but there's a plugin called "Open Folder Context Menus for VS Code" that "Quickly opens a new instance of VS Code from the Explorer". Works pretty well for me.

**Comment by jamietre** [comment]
@alexdresko the problem isn't opening VSCode, it's finding the correct instance among a bunch of already opened ones.

**Comment by saborrie** [comment]
I often have multiple windows of VSCode open. Until I found this issue I'd had a different thought on how to solve it: **setting a workspace/folder colour.**

I commonly have two or more windows open, and sometimes one using the integrated debugging feature for a backend written in NodeJs.  The bottom bar goes orange when debugging, and I've been using that help distinguish which is which.

You can see this in a picture already posted in this thread:

<img src="https://cloud.githubusercontent.com/assets/573056/19929754/4e083920-a0db-11e6-810a-24f161bfcd71.png">

I tend to switch between full-screen windows using the keyboard, rather than hovering on the taskbar to select a window. But even when flicking between full-screen windows, it's not always clear where you've switched to. And while being able to customise the title of the window would give me somewhere to glance at, I think that colourising some part of the window would allow for even quicker discerning of which instance has been focused.

Although the title of this issue is "Allow to configure title bar information" I figured that my suggestion probably belonged here at first rather than in it's own issue, since the other comments here have equated this issue to "Allow to configure easy distinction between open windows". But I can create a new issue if others agree with my suggestion but don't want to lump the two together.

**Comment by chief1ic** [comment]
+1. I frequently work on multiple sub-projects of a large project at the same time. {filename} - {root folder} doesn't help because, for other reasons, all of my root folders are 'src' in VSCode for each of my projects.

**Comment by gilboa23** [comment]
The ability to configure the window title to show `"{rootFolder} - {fileName}"` or even better  just `"{rootFolder}"` instead of `"{fileName} - {rootFolder}"` is critical for us. We're working in a multiple project setup in Windows and it is currently impossible to distinguish open vscode windows on the taskbar.

This seems to be a rather trivial configuration option to implement, but it adds tremendous usability value.

**Comment by bpasero** [comment]
From the discussion I can see the following template variables as being considered useful:
* `{activeFileName}`: e.g. `myFile.txt`
* `{activeFilePath}`: e.g. `/Users/Development/myProject/myFile.txt`
* `{workspaceName}`: e.g. `myProject`
* `{workspacePath}`: e.g. `/Users/Development/myProject`
* `{appName}`: e.g. `VS Code`

(the current git branch is not straight forward to add for layering issues but maybe later)

There is still something not clear to me: We will probably soon support multiple root folders per workspace. In that case `{workspaceName}` or `{workspacePath}` do not make much sense. If I had the folders `myProject` and `otherProject` open in the same window, I would probably expect the title to show me `myProject, otherProject`. In that case, showing the full path of both does not make much sense either. Maybe instead of `{workspaceName}` it should rather be called `{projectName}`. `{workspacePath}` could be `{projectPath}` and be somehow smartly computed from the root folders (however that works...).

**Comment by alexdresko** [comment]
I think of workspaces as more of a conceptual entity - like how I can have a VS .SLN with a name that is completely different from the containing folder. If you're going to allow multiple folders, you're essentially creating something similar to a VS .SLN, right? That workspace name is something we'd want to add to the title bar, as we as, separately, the path to the workspace file. At that point, the folders in the workspace don't really matter. I don't expect VS to show me the path to all of the projects in the .SLN in the title bar. The name and path to the currently open file are important though.

**Comment by bpasero** [comment]
I settled on the following solution: there is a new setting `window.title` (the previous `window.showFullPath` is no longer supported) that allows to configure the title based on some variables.

Variables are declared using `$()` syntax and the following are available:
* `$(activeEditorName)`: e.g. `myFile.txt`
* `$(activeFilePath)`: e.g. `/Users/Development/myProject/myFile.txt`
* `$(rootName)`: e.g. `myProject`
* `$(rootPath)`: e.g. `/Users/Development/myProject`
* `$(appName)`: e.g. `VS Code`
* `$(dirty)`: a dirty indicator if the active editor is dirty
* `$(separator)`: a conditional separator (" - ") that only shows when surrounded by variables with values

Default values on the various OS:

**macOS:**
`$(activeEditorName)$(separator)$(rootName)`

**Linux/Windows:**
`$(dirty)$(activeEditorName)$(separator)$(rootName)$(separator)$(appName)`

**Comment by alexdresko** [comment]
Is this in the latest insiders build?

**Comment by bpasero** [comment]
@alexdresko it will be tomorrow. I did a slight update to let the variables be `${}` instead of `$()`.

Happy for everyones feedback once this is out.

**Comment by jamietre** [comment]
w00t!!!!

![image](https://cloud.githubusercontent.com/assets/573056/22692485/53014d6c-ed0d-11e6-9228-cb7026e58573.png)

**Comment by bpasero** [comment]
Yeah, I can see how a scripted solution would be better, but let's see how far we get with this approach for now.

**Comment by jamietre** [comment]
It would really just be a more universal solution to this kind of thing, generally, but I assume it involves some kind of infrastructure change or new convention, since config is `json` & not `js`. 

Really just threw that out there for long-term consideration. I'm very happy with the way this is working.

**Comment by bpasero** [comment]
Actually I wonder if `${separator}` should be dropped, it seems there is really only one usecase for it: to separate between 2 variables if (and only if) both variables are present:

`${foo}{separator}${bar}`

Why not just automatically add a separator if there is `${foo}${bar}`? Makes it easier to read and gets rid of a complicated and confusing variable.

**Comment by jamietre** [comment]
You might not always want it, e.g. `${dirty}${foo}`

**Comment by bpasero** [comment]
Hm yeah 💩

**Comment by gilboa23** [comment]
Thanks, I've been waiting for this for long time.
Small changes sometime make a big difference :)

**Comment by tommck** [comment]
I really like it.  Thank you for this update.

You *might* want to think about allowing people to add parsing logic (too much?)

In my case, our code is laid out like:

/path/to/project/[Branch Name]

So, using {rootPath} is nice, but I'd like to trim "[Branch Name]" off most of the time.

It might be a bit much to support this though

**Comment by tommck** [comment]
This doesn't seem to be working for me.  I'm not seeing the file name first.. just seeing 

"C:\Path\To\Project - Visual Studio Code - Insiders"

    "window.title": "${dirty}${activeEditorName}${separator}${rootPath}${separator}${appName}",

I restarted the app and opened various files.

Also, does it require a restart to work?  Kind of hard to test what you're doing that way

**Comment by alexdresko** [comment]
Pro tip: You won't see the setting changes when editing the settings file. You have to switch to another file in your workspace to see the changes you've made in the settings. Not sure why that is exactly.

**Comment by tommck** [comment]
I'm still doing that and it's not happening :(

**Comment by tommck** [comment]
The tooltip help for it isn't formatting properly either :)
Should I open a bug?

![image](https://cloud.githubusercontent.com/assets/981957/22977201/af3d753e-f35b-11e6-916f-a11b198d313e.png)

**Comment by tommck** [comment]
Huh.. I think it's actually updating, but taking a LONG time to do so on this machine.  So, I think my problem with the title bar is not a problem.

**Comment by tommck** [comment]
Hmm.. now we just need this feature for the "recent window" list, which just shows the folder name for me, which isn't specific enough (lots of "main" listed in there)

**Comment by bpasero** [comment]
Small update, I added `${activeFilePathRelative}` to be able to show a shorter path to the active opened file if the file is from within the workspace opened. E.g. instead of `/Users/Development/myProject/myFolder/myFile.txt` showing `myFolder/myFile.txt` if `myFolder` is opened. Together with `${rootName}` this allows to show a file path as `myProject/myFolder/myFile.txt` if desired.

**Comment by bpasero** [comment]
Another update after feedback from testing during this week from the team: `${activeFilePath}` and `${activeFilePathRelative}` are working only if a file is opened, making the configuration of the `window.title` hard when non-file editors are opened. 

I decided to revisit the variables around file paths and change them because things are getting complicated when an editor is not showing a file. 

Here is my new solution:
* `${activeEditorLong}`: e.g. /Users/Development/myProject/myFolder/myFile.txt
* `${activeEditorMedium}`: e.g. myFolder/myFile.txt
* `${activeEditorShort}`: e.g. myFile.txt

If there is no long editor label, the medium one will be used. And the same is true for medium and short.

Basically:
* replace `${activeFilePath}` with `${activeEditorLong}`
* replace `${activeFilePathRelative}` with `${activeEditorMedium}`

I hope this works for everyone, should be available in today's insider build.

**Comment by nfour** [comment]
@bpasero is there a way to remove the `[Unsupported]` suffix in the title, when vscode has a modified installation (Using custom CSS)?

**Comment by bpasero** [comment]
No, we strongly advice against using extensions that modify our bits so it is in the best interest to show this not only to you as a user but also to us when we have to investigate bugs which are caused by extensions that patch our bits.

**Comment by VojtechVitek** [comment]
@bpasero `"window.title"` is a neat addition. Can we have something similar for tabs?

![screen shot 2017-03-22 at 2 12 13 pm](https://cloud.githubusercontent.com/assets/139342/24213656/94e2910a-0f09-11e7-805f-3bfa3202a53f.png)

I'd love to see the last directory name as a prefix to the filename on all tabs.

Vim tabs are very smart about tab names btw.

**Comment by bpasero** [comment]
Yeah this was reported already as feature request.

**Comment by nileshp87** [comment]
I made a quick extension for setting the title for the workspace: https://marketplace.visualstudio.com/items?itemName=nilpatel.title

Those who find this thread may find it useful.

**Comment by VojtechVitek** [comment]
@nileshp87 works great, thank you!

![screen shot 2017-04-26 at 7 15 02 pm](https://cloud.githubusercontent.com/assets/139342/25461041/bc162fc2-2ab4-11e7-9990-7107f83d579a.png)
